### ISSUE TITLE
StepLR, MultiStepLR, ExponentialLR and CosineAnnealingLR scheduler wrong lr value

### ISSUE THREAD

**Issue by jasam-sheja** [pytorch/pytorch#20527]
When the StepLR, MultiStepLR, ExponentialLR or CosineAnnealingLR scheduler is called with the same epoch parameter the optimizer value is further reduced even though it's the same epoch

a sample code
```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)
```
now if we call `lr_scheduler.step(epoch=2)` multiple times 
```
for _ in range(10):
    lr_scheduler.step(2)
    print(optimizer.param_groups[0]['lr'])
```
output:
```
>>> 0.0001
>>> 1e-05
>>> 1.0000000000000002e-06
>>> 1.0000000000000002e-07
>>> 1.0000000000000004e-08
>>> 1.0000000000000005e-09
>>> 1.0000000000000006e-10
>>> 1.0000000000000006e-11
>>> 1.0000000000000006e-12
>>> 1.0000000000000007e-13
```

Even if such use-case is bizarre, this is extremely unexpected.
This is happens using PyTorch version 1.1.0 but not 1.0.1. Because pull(#14010) redefined StepLR, MultiStepLR, ExponentialLR and CosineAnnealingLR to directly use the learning rate variable of the optimizer rather than using `base_lr` defined by `_LRScheduler`. This was in order to support multiple simultaneous schedulers (#13022)!

**Comment by ezyang** [comment]
cc @chandlerzuo

**Comment by ezyang** [comment]
@jasam-sheja, if I understand you correctly, there is not a math formula problem; you just want to not have the optimizer value reduce if it is called with the same epoch value?

**Comment by jasam-sheja** [comment]
@ezyang, I don't think there is a math formula problem. I just expect the schedule to be consistent with the epoch, as if there's a time table specifying the learning rate at each epoch. Otherwise the `epoch` parameter is misleading.
To answer your question; yes I don't want it to reduce or increase if it is called with the same epoch value.

**Comment by jafermarq** [comment]
is there a quick fix for this?

**Comment by ezyang** [comment]
@jafermarq I'm going to put up a PR that reverts the breaking commit to start. I'm also planning on making a fix, though that might take longer.

**Comment by ezyang** [comment]
Totally untested revert PR at #21375. Waiting for tests.

**Comment by soumith** [comment]
ummm yea this is a wtfbug, sorry about that! Thanks @ezyang for reverting it

**Comment by ezyang** [comment]
I'm still planning to revert. But let me explain why this is "expected behavior" under the patch @chandlerzuo wrote.

When we landed #14010, the intended API was to let you do something like this:

```
for epoch in range(1000):
    lr_scheduler1.step(epoch)
    lr_scheduler2.step(epoch)
    train()
    validate()
```

Because `lr_scheduler1` updates the rates, `lr_scheduler2` is able to do its update with respect to `lr_scheduler1`'s update. Makes sense.

But what if you call `lr_scheduler1` twice, with the same epoch?

```
for epoch in range(1000):
    lr_scheduler1.step(epoch)
    lr_scheduler1.step(epoch)
    train()
    validate()
```

That's precisely the bug reported here. But we've done "what we said we were going to do": we have composed `lr_scheduler1` with itself, which is **not** an idempotent operation (and probably not what you wanted, but hey, you asked for it.)

We're going to revert because the "calling scheduler on the same epoch" seems like a more important invariant to preserve. We'll probably implement compositional LRSchedulers a different way, probably by literally having a ComposeLRScheduler. Sorry about the inconvenience in the meantime.

**Comment by ezyang** [comment]
@jafermarq If you apply #21463 this should fix it for you.

**Comment by fmassa** [comment]
Let me know if you revert the composable LR Schedulers (which would be btw a breaking change wrt 1.1 API).

I've used the composable LR schedulers in the reference training scripts in torchvision, so I'd like to update it / find a solution if this gets reverted

**Comment by vincentqb** [comment]
Do we have examples of why the users are passing the epoch number as a parameter? Since the schedulers are stateful, I would have expected the schedulers to know how many times they were called, if relevant.

If the goal is to recover prior learning rate, I would expect to have a separate interface to do that, which may return the history of learning rates generated by the scheduler.

If the goal is to skip a few epochs in the scheduler, a parameter could simply be provided say "skip_num_epochs" to jump ahead that number of epochs in the scheduler.

Thoughts?

**Comment by vincentqb** [comment]
Only schedulers depending only on epoch number can have the epoch parameter as an option, and so this parameter should not be expected from a general scheduler. (ReduceLROnPlateau is an example of scheduler that should not have it.) We are planning on reinstating #14010 with the following modifications: when the user inputs the epoch parameter, a warning is shown, and the closed form of the scheduler is used when available.

**Comment by vincentqb** [comment]
Change described in previous comment implemented in #21800

**Comment by vincentqb** [comment]
#24352

**Comment by muammar** [comment]
I also am having problems using `MultiStepLR`,  these are the parameters  I use:

```
scheduler = ("MultiStepLR", {"milestones": [30, 50, 80], "gamma": 0.1})
```

I am getting the current learning rate at each epoch using this function

```
def get_lr(optimizer):
    """Get current learning rate
    
    Parameters
    ----------
    optimizer : obj
        An optimizer object.
    
    Returns
    -------
    lr
        Current learning rate.
    """
    for param_group in optimizer.param_groups:
        return param_group['lr']
```

What I do to get its magnitude is to call  `scheduler.step(loss)` and  `get_lr(optimizer)`. The output of this looks this way: 
 
```
Epoch 1 lr 0.1
Epoch 2 lr 0.00010000000000000003
Epoch 3 lr 0.0010000000000000002
Epoch 4 lr 0.0010000000000000002
Epoch 5 lr 0.0010000000000000002
Epoch 6 lr 0.0010000000000000002
Epoch 7 lr 0.0010000000000000002
Epoch 8 lr 0.010000000000000002
```
As you can see from above, its behavior is very erratic.  How can this be fixed?

**Comment by ezyang** [comment]
@muammar What version of PyTorch are you on?

**Comment by muammar** [comment]
> @muammar What version of PyTorch are you on?

@ezyang This is the version I am using: 

```
In [2]: import torch                                                            

In [3]: torch.__version__                                                       
Out[3]: '1.3.1'
```

**Comment by vincentqb** [comment]
The original issue has been addressed in #26423. In particular, there is now a warning against calling schedulers with constant epoch.

```
>>> import torch
>>> torch.__version__
'1.3.1'
>>> import torch.optim as optim
>>> from torch import nn
>>> conv = nn.Conv2d(3,3,3)
>>> optimizer = optim.Adam(conv.parameters()) 
>>> 
>>> lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)
>>> for _ in range(10):
...     lr_scheduler.step(2)
...     print(optimizer.param_groups[0]['lr'])
... 
/Users/username/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/optim/lr_scheduler.py:100: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule.See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  "https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate", UserWarning)
0.0001
0.0001
0.0001
0.0001
0.0001
0.0001
0.0001
0.0001
0.0001
0.0001
>>> 
```

Can you open a new issue with example code? I cannot reproduce.

```
>>> import torch
>>> torch.__version__
'1.3.1'
>>> import torch.optim as optim
>>> from torch import nn
>>> conv = nn.Conv2d(3,3,3)
>>> optimizer = optim.Adam(conv.parameters()) 
>>> 
>>> lr_scheduler = optim.lr_scheduler.MultiStepLR(optimizer, milestones=[30, 50, 80], gamma=0.1)
>>> for _ in range(10):
...     lr_scheduler.step(2)
...     print(optimizer.param_groups[0]['lr'])
... 
/Users/username/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/optim/lr_scheduler.py:100: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule.See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  "https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate", UserWarning)
0.001
0.001
0.001
0.001
0.001
0.001
0.001
0.001
0.001
0.001
```

I will close the issue for now. Please feel free to re-open or creating a new issue with example code if this continues.

**Comment by mugheesahmad** [comment]
I ran this code on colab and the output is not consistent.

import torch
print("pytorch version",torch.__version__) 
import torch.nn as nn
model = nn.Linear(1, 1) # 'Net' is a simple MLP
optimizer = torch.optim.SGD(model.parameters(), lr=0.1)
schedular = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones = [4,7], gamma=0.1)

print('Initial LR : {0:.8f}'.format(schedular.get_lr()[0]))
for e in range(8):
    schedular.step()
    print('Epoch {0}, LR: {1:.8f}, opt LR {2:.8f}'.format(e, schedular.get_lr()[0],
          optimizer.param_groups[0]['lr']))


pytorch version 1.4.0
Initial LR : 0.10000000
Epoch 0, LR: 0.10000000, opt LR 0.10000000
Epoch 1, LR: 0.10000000, opt LR 0.10000000
Epoch 2, LR: 0.10000000, opt LR 0.10000000
**Epoch 3, LR: 0.00100000, opt LR 0.01000000**
Epoch 4, LR: 0.01000000, opt LR 0.01000000
Epoch 5, LR: 0.01000000, opt LR 0.01000000
**Epoch 6, LR: 0.00010000, opt LR 0.00100000**
Epoch 7, LR: 0.00100000, opt LR 0.00100000

**Comment by vincentqb** [comment]
Thanks for the example. As mentioned in #26423, `get_lr` should be replaced by `get_last_lr`, see below.

```python
import torch
print("pytorch version",torch.__version__) 
import torch.nn as nn
model = nn.Linear(1, 1) # 'Net' is a simple MLP
optimizer = torch.optim.SGD(model.parameters(), lr=0.1)
schedular = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones = [4,7], gamma=0.1)

print('Initial LR : {0:.8f}'.format(schedular.get_lr()[0]))
for e in range(8):
  optimizer.step()
  schedular.step()
  print('Epoch {0}, LR: {1:.8f}, opt LR {2:.8f}'.format(e, schedular.get_last_lr()[0],
          optimizer.param_groups[0]['lr']))
```

Since #26423 has been merged, I will close this issue. Please feel free to re-open if the issue persists.
